index.bs

```diff
@@ -734,7 +753,271 @@ Each {{WebRequestEvent}} has:
+
+  1. Let |handlers| be the result of calling [=lookup registered WebRequest
+      handler configs=] given "beforeRequest", and |request|.
+
+  1. For each |handler| in |handlers|:
+
+      1. Let |details| be the result of calling [=create a
+          WebRequestEventDetails object=] given |request|.
+
+      1. If |request|'s [=request/body=] is not null, then:
+
+          1. Let |request_body| be a new {{RequestBody}}.
+
+          1. Let |body| be |request|'s [=request/body=].
+
+              <span class=XXX>TODO: serialize [=body/stream=] if present.</span>
+
+          1. Switch on |body|'s [=body/source=]:
+
+              : [=byte sequence=]
+              :: [=list/Append=] a new {{UploadData}} with {{UploadData/bytes}}
+                  equal to the serialized |body|'s [=body/source=] to
+                  |request_body|'s {{RequestBody/raw}}.
+
+              : {{Blob}}
+              :: [=list/Append=] a new {{UploadData}} with {{UploadData/bytes}}
+                  equal to the serialized |body|'s [=body/source=] to
+                  |request_body|'s {{RequestBody/raw}}.
+
+              : {{FormData}}
+              ::  1. Let |formData| be a new {{/object}}.
+
+                  1. [=list/For each=] |entry| in |body|'s [=body/source=]'s
+                      [=FormData/entry list=]:
+
+                      1. Switch on |entry|[1]:
+
+                      : {{File}}
+                      :: [=list/Append=] a new {{UploadData}} with
+                          {{UploadData/file}} equal to |entry|[1]'s
+                          {{File/name}} to |request_body|'s
+                          {{RequestBody/raw}}.
+
+                      : {{USVString}}
+                      ::  1. If |formData|[|entry|[0]] does not exist, then
+                              set |formData|[|entry|[0]] equal to an empty
+                              [=list=].
+
+                          1. [=list/Append=] |entry|[1] to
+                              |formData|[|entry|[0]].
+
+
+                  1. Set |details|'s {{RequestBody/formData}} to |formData|.
+
+          1. Set |details|'s {{WebRequestBeforeRequestDetails/requestBody}}
+              to |request_body|.
+
+      1. If |handler|'s [=WebRequest handler config/blocking=] flag is true,
+          then:
+
+          1. Let |result| be the result of calling |handler|[[=WebRequest
+              handler config/handler=]] given |details|.
+
+          1. If any [=map/key=] in |result| is not [=list/contained=] in the set
+              « "cancel", "redirect" », then [=throw=] a {{TypeError}}.
+
+          1. Return |result|.
+
+      1. Call |handler|[[=WebRequest handler config/handler=]] given |details|
+          [=in parallel=].
+
+      1. Return null.
+
+</div>
+
+<div algorithm>
+  To <dfn>create a {{WebRequestEventDetails}} object</dfn> given a [=request=]
+  |request|, run the following steps:
+
+  1. Let |environment settings object| be |request|'s [=request/client=].
```

Comment:
|environmentSettingsObject| ? Or maybe there's enough cases of this in existing specs elsewhere that I might be convinced that leaving it as-is is better.